apps/web/components/dashboard-components/project-details/custom-llm-parameters-editor.tsx

```diff
@@ -0,0 +1,650 @@
+        Object.keys(existingParams).length > 0 ? existingParams : null;
+    }
+
+    updateProject.mutate({
+      projectId: project.id,
+      customLlmParameters,
+    });
+  };
+
+  const handleCancel = () => {
+    // Reset to original values
+    if (project?.customLlmParameters) {
+      // Use selectedProvider if provided, otherwise fall back to project's default provider
+      const currentProvider =
+        selectedProvider || project.defaultLlmProviderName;
+      // Use selectedModel if provided, otherwise fall back to project's default model
+      const currentModel = selectedModel || project.defaultLlmModelName;
+
+      if (currentProvider && currentModel) {
+        const providerParams = project.customLlmParameters[currentProvider];
+        const modelParams = providerParams?.[currentModel];
+
+        if (modelParams && typeof modelParams === "object") {
+          const entries = Object.entries(modelParams).map(([key, value]) => ({
+            key,
+            value: String(value),
+            type:
+              typeof value === "boolean"
+                ? "boolean"
+                : typeof value === "number"
+                  ? "number"
+                  : "string",
+          })) as ParameterEntry[];
+          setParameters(entries);
+        } else {
+          setParameters([]);
+        }
+      } else {
+        setParameters([]);
+      }
+    } else {
+      setParameters([]);
+    }
+    setIsEditing(false);
+    setHasUnsavedChanges(false);
+  };
+
+  const addParameter = () => {
+    setParameters([...parameters, { key: "", value: "", type: "string" }]);
+    setHasUnsavedChanges(true);
+  };
+
+  const removeParameter = (index: number) => {
+    setParameters(parameters.filter((_, i) => i !== index));
+    setHasUnsavedChanges(true);
+  };
+
+  const updateParameter = (
+    index: number,
+    field: keyof ParameterEntry,
+    value: string,
+  ) => {
+    const newParameters = [...parameters];
+    newParameters[index] = { ...newParameters[index], [field]: value };
+    setParameters(newParameters);
+    setHasUnsavedChanges(true);
+  };
+
+  const applySuggestion = (suggestion: { key: string; type: string }) => {
+    // Check if parameter already exists
+    const existingIndex = parameters.findIndex((p) => p.key === suggestion.key);
+    if (existingIndex >= 0) {
+      toast({
+        description: `Parameter "${suggestion.key}" already exists`,
+      });
+      return;
+    }
+
+    // Add the suggested parameter with default value
+    const defaultValue =
+      suggestion.type === "boolean"
+        ? "false"
+        : suggestion.type === "number"
+          ? "0"
+          : "";
+    setParameters([
+      ...parameters,
+      {
+        key: suggestion.key,
+        value: defaultValue,
+        type: suggestion.type as ParameterEntry["type"],
+      },
+    ]);
+    setHasUnsavedChanges(true);
+  };
+
+  // If project is undefined, show a loading state
+  if (!project) {
```

Comment:
I find this pattern leads to a lot of excess flashing/animation on the screen as the DOM nodes differ slightly between loading states - especially given that you have a motion.div in one case (below) but not here. Is there any way you can just always return the same dom tree but leave some of the fields blank while the project loads?